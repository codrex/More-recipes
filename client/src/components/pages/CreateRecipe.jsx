import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import RecipeEditor from './RecipeEditor';
import { resetSuccess } from '../../actions/ajaxActions';
import { RECIPE_ADDED } from '../../constants';
import {
  createRecipe,
  getRecipe,
  resetRecipe
} from '../../actions/recipeActions';

/**
 * CreateRecipe
 */
class CreateRecipe extends React.Component {
  /**
   * constructor
   * @return {undefined}
   * @param {Object} props
   */
  constructor(props) {
    super(props);
    props.resetRecipe();
  }

  /**
   * @returns {undefined}
   * @param {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    const {
      message,
      history,
      resetSuccessMessage
    } = nextProps;

    if (message === RECIPE_ADDED) {
      resetSuccessMessage();
      history.push(`/recipe/${nextProps.recipe.id}`);
    }
  }

  /**
   * post recipe button click event handler
   * @return {undefined}
   */
  postRecipeClicked = () => {
    const { recipe } = this.props;
    this.props.createRecipe(recipe, RECIPE_ADDED);
  }

  /**
   * render
   * @return {React} - react component
   */
  render() {
    const { loading } = this.props;

    return (
      <RecipeEditor
        handleButtonClick={this.postRecipeClicked}
        submitButtonText="post recipe"
        loading={loading}
        hasRender
      />
    );
  }
}

CreateRecipe.propTypes = {
  recipe: PropTypes.objectOf(PropTypes.shape).isRequired,
  history: PropTypes.objectOf(PropTypes.shape).isRequired,
  resetRecipe: PropTypes.func.isRequired,
  createRecipe: PropTypes.func.isRequired,
  resetSuccessMessage: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  message: PropTypes.string,
};

CreateRecipe.defaultProps = {
  message: '',
};

const mapStateToProps = state => (
  {
    recipe: state.recipe,
    loading: state.networkRequest.loading,
    message: state.networkRequest.msg,
    statusCode: state.currentStatusCode
  }
);

export { CreateRecipe as PureCreateRecipe };
export default connect(mapStateToProps, {
  resetRecipe,
  createRecipe,
  getRecipe,
  resetSuccessMessage: resetSuccess
})(CreateRecipe);
