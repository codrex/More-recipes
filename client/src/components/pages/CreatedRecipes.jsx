import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getCreatedRecipes } from '../../actions/recipeActions';
import RecipesDisplay from './RecipesDisplay';

/**
   * @return {React} Favourite Recipe component
   * @param {object} props
   */
const CreatedRecipes = props => (
  <RecipesDisplay
    title="Created recipes"
    getRecipes={props.getCreatedRecipes}
    recipes={props.createdRecipes}
    {...props}
  />
);

CreatedRecipes.propTypes = {
  getCreatedRecipes: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  createdRecipes: state.recipes
});

export { CreatedRecipes as PureCreatedRecipes };
export default connect(mapStateToProps, {
  getCreatedRecipes
})(CreatedRecipes);
