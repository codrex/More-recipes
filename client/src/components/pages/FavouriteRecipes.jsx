import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getFavouriteRecipes } from '../../actions/recipeActions';
import RecipesDisplay from './RecipesDisplay';

/**
 * @return {React} Favourite Recipe component
 * @param {object} props
 */
const FavouriteRecipes = props => (
  <RecipesDisplay
    title="Favourite recipes"
    getRecipes={props.getFavouriteRecipes}
    recipes={props.favoriteRecipes}
    {...props}
  />
);
FavouriteRecipes.propTypes = {
  getFavouriteRecipes: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  favoriteRecipes: state.favoriteRecipes
});
export { FavouriteRecipes as PureFavoriteRecipes };
export default connect(mapStateToProps, {
  getFavouriteRecipes
})(FavouriteRecipes);
