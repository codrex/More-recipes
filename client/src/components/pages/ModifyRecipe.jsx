import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import toastr from 'toastr';
import RecipeEditor from './RecipeEditor';
import { modifyRecipe, getRecipe } from '../../actions/recipeActions';
import { resetSuccess } from '../../actions/ajaxActions';
import { RECIPE_MODIFIED } from '../../constants';
import toastrConfig from '../../toastr/config';

/**
 * ModifyRecipe
 */
class ModifyRecipe extends React.Component {
  /**
   * constructor
   * @return {undefined}
   * @param {Object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      hasNotFound: false,
    };
    this.recipeId = parseInt(this.props.match.params.id, 10);
  }

  /**
   * @returns {undefined}
   */
  componentDidMount() {
    if (!isNaN(this.recipeId)) {
      this.props.getRecipe(this.recipeId);
    } else if (!this.state.hasNotFound) {
      // eslint-disable-next-line
      this.setState({ hasNotFound: true });
      toastr.error('Sorry, recipe not found', 'Error', toastrConfig);
    }
  }

  /**
   * @returns {undefined}
   * @param {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    const {
      message,
      statusCode,
      history,
      resetSuccessMessage
    } = nextProps;

    if (message === RECIPE_MODIFIED) {
      resetSuccessMessage();
      history.push(`/recipe/${nextProps.recipe.id}`);
    }

    if (statusCode === 404 && !this.state.hasNotFound) {
      this.setState({ hasNotFound: true });
    }
  }

  /**
   * @return {undefined}
   */
  modifyButtonClicked = () => {
    const { recipe } = this.props;
    this.props.modifyRecipe(recipe, RECIPE_MODIFIED);
  }

  /**
   * render
   * @return {React} - react component
   */
  render() {
    const { hasNotFound } = this.state;
    const { loading, recipe } = this.props;
    const hasDelayedRendering = recipe.id === this.recipeId;

    return (
      <RecipeEditor
        hasNotFound={hasNotFound}
        handleButtonClick={this.modifyButtonClicked}
        hasRender={hasDelayedRendering}
        submitButtonText="modify recipe"
        loading={loading}
      />
    );
  }
}

ModifyRecipe.propTypes = {
  recipe: PropTypes.objectOf(PropTypes.shape).isRequired,
  history: PropTypes.objectOf(PropTypes.shape).isRequired,
  match: PropTypes.objectOf(PropTypes.shape).isRequired,
  getRecipe: PropTypes.func.isRequired,
  modifyRecipe: PropTypes.func.isRequired,
  resetSuccessMessage: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  message: PropTypes.string,
  statusCode: PropTypes.number.isRequired
};

ModifyRecipe.defaultProps = {
  message: '',
};

const mapStateToProps = state => (
  {
    recipe: state.recipe,
    loading: state.networkRequest.loading,
    message: state.networkRequest.msg,
    statusCode: state.currentStatusCode
  }
);

export { ModifyRecipe as PureModifyRecipe };
export default connect(mapStateToProps, {
  modifyRecipe,
  getRecipe,
  resetSuccessMessage: resetSuccess
})(ModifyRecipe);
