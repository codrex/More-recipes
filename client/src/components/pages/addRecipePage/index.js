import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Icon from '../../common/icon/icon';
import AddDirections from './addDirections/addDirections';
import AddIngredients from './addIngredients/addIngredients';
import RecipeNameAndCategory from './recipeNameAndCategory/recipeNameAndCategory';
import { createRecipe, updateAllRecipeField } from '../../../actions/recipeActions';
import { ajaxRedirect } from '../../../actions/ajaxActions';
import './addRecipe.scss';

/**
 * Add recipe component
 */
class AddRecipes extends React.Component {

  /**
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.saveRecipe = this.saveRecipe.bind(this);
  }
/**
 *
 * @param {object} nextProps
 * @return {bool} true or false
 */
  shouldComponentUpdate(nextProps) {
    if (nextProps.redirectUrl === '/') {
      this.props.actions.redirect('/recipe/create');
      this.props.history.push(nextProps.redirectUrl);
      return false;
    }
    if (nextProps.recipe !== this.props.recipe) {
      this.props.actions.updateLocalRecipe(nextProps.recipe);
      return true;
    }
    return true;
  }

  /**
   * @return {undefined}
   * @param {Object} values
   */
  saveRecipe(values) {
    this.props.actions.updateStore(values);
  }
  /**
   * @return {undefined}
   */
  render() {
    return (
      <div className="container-fluid no-padding ">
        <div className="row d-flex">
          <div className="col-xs-9 col-sm-9 col-md-10 col-lg-4 name-category no-padding">
            <h4 className="lead items-header-text">
              Add recipe
              <Icon
                iconClass="fa fa-plus"
                className="float-right"
                handleClick={this.newItem}
              />

            </h4>
            <div className="col-12">
              <RecipeNameAndCategory createRecipe={createRecipe} />
            </div>
          </div>
          <div className="col-xs-9 col-sm-9 col-md-10 col-lg-4 d-flex ingredients no-padding">
            <AddIngredients />
          </div>
          <div className="col-xs-9 col-sm-9 col-md-10 col-lg-4 d-flex directions no-padding">
            <AddDirections />
          </div>
        </div>
      </div>
    );
  }
}

AddRecipes.propTypes = {
  actions: PropTypes.object,
  loading: PropTypes.bool,
  redirectUrl: PropTypes.string,
  history: PropTypes.object,
  redirect: PropTypes.func,
  recipe: PropTypes.object,

};

const mapStateToProps = (state) => (
  {
    loading: state.ajaxCall > 0,
    redirectUrl: state.redirectUrl,
    recipe: state.recipe,
  }
);
const mapDispatchToProps = (dispatch) => (
  {
    actions: {
      redirect: bindActionCreators(ajaxRedirect, dispatch),
      updateLocalRecipe: bindActionCreators(updateAllRecipeField, dispatch)
    }
  }
);

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipes);
