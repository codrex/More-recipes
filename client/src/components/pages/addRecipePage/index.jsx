import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import TopBar from '../../common/topbar/topbar';
import AddDirections from './addDirections/addDirections';
import AddIngredients from './addIngredients/addIngredients';
import RecipeNameAndCategory from './recipeNameAndCategory/recipeNameAndCategory';
import {
  updateAllRecipeField,
  recipeToModify,
  getRecipe,
  onNewRecipe
} from '../../../actions/recipeActions';
import { ajaxRedirect } from '../../../actions/ajaxActions';

/**
 * Add recipe component
 */
class AddRecipes extends React.Component {
  /**
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.create = props.match.path === '/recipe/create';
    this.CREATE_RECIPE_SUCCESS = 'Recipe added successfully';
    this.MODIFY_RECIPE_SUCCESS = 'Recipe modified successfully';
  }

  /**
 * @return {undefined}
 */
  componentDidMount() {
    if (!this.create) {
      const id = parseInt(this.props.match.params.id, 10);
      if (this.props.recipes) {
        const recipe = this.props.recipes.find(element => element.id === id);
        this.props.actions.recipeToModify(recipe);
      } else {
        this.props.actions.getRecipe(id);
      }
    } else {
      this.props.actions.newRecipe();
    }
  }
  /**
   * @return {undefined}
   * @param {Object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.success.success === this.CREATE_RECIPE_SUCCESS) {
      this.props.actions.newRecipe();
      this.props.actions.clearSuccessMsg({
        success: ''
      });
    }
  }

  /**
 *
 * @param {object} nextProps
 * @return {bool} true or false
 */
  shouldComponentUpdate(nextProps) {
    if (nextProps.redirectUrl === '/') {
      this.props.actions.redirect('/recipe/create');
      this.props.history.push(nextProps.redirectUrl);
      return false;
    }
    if (nextProps.recipe !== this.props.recipe) {
      this.props.actions.updateLocalRecipe(nextProps.recipe);
      return true;
    }
    return true;
  }

  /**
   * @return {undefined}
   */
  render() {
    return (
      <div
        className="container"
        style={{
          maxWidth: 'unset'
        }}
      >
        <div className="row">
          <TopBar title={this.create ? 'Add recipe' : 'Modify recipe'} />
        </div>
        <div
          className="row d-flex justify-content-between align-items-start"
          style={{
            margin: '0.5em'
          }}
        >
          <div className="col-xs-11 col-sm-11 col-md-10 col-lg-4 name-category no-padding">
            <div className="col-12">
              <RecipeNameAndCategory
                postRecipe={this.create}
                loading={this.props.loading}
                create={this.CREATE_RECIPE_SUCCESS}
                modify={this.MODIFY_RECIPE_SUCCESS}
              />
            </div>
          </div>
          <div className="col-xs-11 col-sm-11 col-md-10 col-lg-4 d-flex ingredients no-padding">
            <AddIngredients />
          </div>
          <div className="col-xs-11 col-sm-11 col-md-10 col-lg-4 d-flex directions no-padding">
            <AddDirections />
          </div>
        </div>
      </div>
    );
  }
}

AddRecipes.propTypes = {
  actions: PropTypes.object,
  loading: PropTypes.bool,
  redirectUrl: PropTypes.string,
  history: PropTypes.object,
  redirect: PropTypes.func,
  recipe: PropTypes.object,
  recipes: PropTypes.array,
  match: PropTypes.object
};

const mapStateToProps = state => ({
  loading: state.ajaxCall > 0,
  redirectUrl: state.redirectUrl,
  // recipes: state.user.createdRecipes,
  success: state.ajaxSuccess
});
const mapDispatchToProps = dispatch => ({
  actions: {
    redirect: bindActionCreators(ajaxRedirect, dispatch),
    updateLocalRecipe: bindActionCreators(updateAllRecipeField, dispatch),
    recipeToModify: bindActionCreators(recipeToModify, dispatch),
    getRecipe: bindActionCreators(getRecipe, dispatch),
    newRecipe: bindActionCreators(onNewRecipe, dispatch),
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(AddRecipes);
