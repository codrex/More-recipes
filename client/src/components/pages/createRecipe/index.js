import CreateRecipe from './createRecipe';

export default CreateRecipe;
