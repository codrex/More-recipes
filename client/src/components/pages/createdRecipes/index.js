import CreatedRecipes from './createdRecipes';

export default CreatedRecipes;
