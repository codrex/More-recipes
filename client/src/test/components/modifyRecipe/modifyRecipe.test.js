import React from 'react';
import toJson from 'enzyme-to-json';
import { PureModifyRecipe } from '../../../components/pages/ModifyRecipe';

const props = {
  match: {
    params: {
      id: 1
    }
  },
  message: 'recipe modified',
  getRecipe: jest.fn(),
  modifyRecipe: jest.fn(),
  resetSuccessMessage: jest.fn(),
  loading: false,
  statusCode: 200,
  recipe: {
    name: 'capachino',
    category: 'drinks',
    views: 1,
    upVotes: 1,
    downVotes: 1,
    id: 1,
    image: {},
  },
  user: {
    id: 1,
    favRecipes: [{
      id: 1,
    }],
  },
  history: { push: jest.fn() },
};

describe('Modify recipe page component ', () => {
  test('render as expected ', () => {
    const wrapper = shallow(<PureModifyRecipe {...props} />);
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
    expect(tree.props.hasNotFound).toBe(false);
    expect(tree).toBeInstanceOf(Object);
  });
  test('render as expected when recipe was not found ', () => {
    const wrapper = shallow(<PureModifyRecipe {...props} />);
    wrapper.setProps({ statusCode: 404 });
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
    expect(tree.props.hasNotFound).toBe(true);
    expect(tree).toBeInstanceOf(Object);
  });
  test('render as expected when recipe id is not a number', () => {
    const wrapper = shallow(<PureModifyRecipe {...{
      ...props,
      match: {
        params: {
          id: 'fkfmf'
        }
      }, }}
    />);
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
    expect(tree.props.hasNotFound).toBe(true);
    expect(tree).toBeInstanceOf(Object);
  });
});
