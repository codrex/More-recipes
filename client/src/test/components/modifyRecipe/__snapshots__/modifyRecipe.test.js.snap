// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modify recipe page component  render as expected  1`] = `
<RecipeEditor
  handleButtonClick={[Function]}
  hasNotFound={false}
  hasRender={true}
  loading={false}
  message={[Function]}
  statusCode={-1}
  submitButtonText="modify recipe"
/>
`;

exports[`Modify recipe page component  render as expected when recipe id is not a number 1`] = `
<RecipeEditor
  handleButtonClick={[Function]}
  hasNotFound={true}
  hasRender={false}
  loading={false}
  message={[Function]}
  statusCode={-1}
  submitButtonText="modify recipe"
/>
`;

exports[`Modify recipe page component  render as expected when recipe was not found  1`] = `
<RecipeEditor
  handleButtonClick={[Function]}
  hasNotFound={true}
  hasRender={true}
  loading={false}
  message={[Function]}
  statusCode={-1}
  submitButtonText="modify recipe"
/>
`;
